### PR TITLE
Adding support for serverless, reacting to api changes, adding tests 

### LINKED ISSUE #91: Serverless support
When attempting to retrieve the index using the GetIndex method by providing the indexName of a serverless index, an HttpRequestException is encountered. The exception message is as follows: `System.Net.Http.HttpRequestException: 'Response status code does not indicate success: 404 (Not Found).'`

Steps to Reproduce:
Call the GetIndex method on a serverless index using the corresponding indexName.
Observe the raised HttpRequestException with the mentioned status code (404).

### LINKED ISSUE #107: Move PineconeClient to global API for managing index operations
Per blog entry: https://www.pinecone.io/blog/global-api/ the API to manage index operations in form of host / operation is now deprecated. They recommend moving to the new global API ASAP (i.e. before they EOL the old one)

### PR DESCRIPTION
**Pull request by maumar**
- Adding support for serverless indexes
- Moving to new global API (also list indexes and list collections now return more data)
- Adding tests (local testing story is not available so using free option - in order to run tests you need to add your API key to the user secrets - https://learn.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-8.0&tabs=windows#set-a-secret)
- Minor API fixes to more accurately reflect the API spec

Fixes #91
Fixes #107

Lot of breaking changes here - there are significant differences between old and new APIs, also pod vs serverless Tests run against both pod and serverless using the free plan. Functionality not available in this plan has not been tested.

NOT TESTED (requires paid plan):
- create collection from index,
- restore index from collection,
- list collections (when there is something in them),
- pod-based index modifications (upscaling, downscaling)
- delete vectors based on filter

### CODE REVIEW

**Comment by maumar** [comment]
still work in progress - I focused on serverless and that works pretty well, but I have no way currently to test if/how much is broken for the pod based indexes. Since the latest update gcp-starter has been broken.

**Comment by neon-sunset** [comment]
Thank you for the contribution!

> there are significant differences between old and new APIs, also pod vs serverless Tests are mainly using serverless, since at the time of writing free pod based offering is broken (indexes won't initialize)

I had initial sketch of porting to the "new API", that is, the way it looked at the moment of introduction of serverless but ran into similar issues with only the old way working with pods and only the new way working with serverless, as well as billing issues. It was also possible in the past to get plain OpenAPI schema from the documentation page but they changed the way it worked and I could no longer manage to extract it or find in the published official SDKs at the time.

That and switching over from a side project tool I wrote that used to rely on Pinecone (hence no tests - the tool pretty much exercised all paths so running sample workloads gave high coverage immediately) led me to ultimately putting this project on the back burner.

Definitely happy to see interest still, if there is anything I can help with to get this done - please let me know.

**Review by spppaul — APPROVED** [review]
I need this PR to use this client on serverless at all, only alternative is raw REST requests

**Comment by neon-sunset** [comment]
@maumar Thank you for working on this! Let me know if I can review this (and look at API and etc.) in full.

On tests dependent on API key - I think it's ok for them to fail for the time being, I'll look into getting a separate account for test environment.

File: src/PineconeClient.cs
Comment (neon-sunset) [inline src/PineconeClient.cs:46]: Can the users now choose the environment of pod-based indexes with free accounts? (if pod-based indexes are still a thing for those)

It was mostly a shorthand overload given in most situations only those three arguments were needed. It's not so shorthand anymore so you may want to either add sensible defaults to arguments (or omit them so the create request leaves them unspecified if allowed) or just remove it altogether if making it work is unfeasible/unsound.

File: src/Types/IndexTypes.cs
Comment (neon-sunset) [inline src/Types/IndexTypes.cs:6]: Perhaps you meant to mark this `internal` and place it in `Rest/Types.cs`? It's not a part of `PineconeClient` or `Index<T>` API (`ListIndexes` returns `IndexDetails[]`).

File: src/PineconeClient.cs
Comment (maumar) [reply src/PineconeClient.cs:46]: For "gcp-starter" it doesn't matter what pod type, number of pods, etc. you specify. As for the environment - anything other than "gcp-starter" will fail on free account. Sensible defaults (like "p1.x1") seem to be a good idea and should be easy to make it work across the board. However, my preference would be to leave the environment variable without the default, rather than default to "gcp-starter" - for better discoverability. Was planning to add xml comments in a bit, so could add a note about "gcp-starter" for a free user. What do you think?

File: src/Types/IndexTypes.cs
Comment (maumar) [reply src/Types/IndexTypes.cs:6]: Sounds good! yes, I wasn't sure about this one

**Comment by maumar** [comment]
@neon-sunset I am close to being done. Probably good for a thorough review tomorrow, but will ping you again after I push the final changes.

File: src/Types/IndexTypes.cs
Comment (maumar) [inline src/Types/IndexTypes.cs:152]: `uint?` rather than 'long?' for pods, replicas and shards?

Comment (neon-sunset) [reply src/Types/IndexTypes.cs:152]: Makes sense, it's unlikely anyone would have more than 4B pods, unless we're in a *really* fast AI takeoff scenario.

File: src/PineconeClient.cs
Comment (neon-sunset) [inline src/PineconeClient.cs:81]: Thanks, this look reasonable!

**Comment by spppaul** [comment]
Do you guys have an ETA on when this will make it into nuget? I have this branch running locally and it works great.

**Comment by neon-sunset** [comment]
@spppaul this is a community-maintained library (and I'm happy this is now actually a community, with Maumar's and Andrii's contributions!) licensed under MIT.

As with many other FOSS projects, the releases and features delivered are the result of work done by contributors in their spare time and are not subject to SLA or specific ETAs that can be committed to. This PR will likely be merged "soon(tm)" but there is no guarantee or promise to be made as this is not a commercial project bound by a contract.

**Comment by neon-sunset** [comment]
@maumar it is failing as the secret is not accessible as GH prevents its leakage for PRs originating from users without secrets write access. It should work once merged.

When ran locally, it appears that currently index is being concurrently deleted from multiple locations. I added a dirty-ish workaround for this to just ignore the exceptions if Pinecone responded with `Index not found`.

**Comment by maumar** [comment]
@neon-sunset thanks for the review and code fixes! I have just pushed in the final set of changes, so from my perspective the PR should be good to go. 

Wrt test complexity - some of it stems from the fact that Pinecone doesn't have a (good) local testing story. My understanding is that it is an outlier - pretty much any other vector DB has a test container available: https://testcontainers.com/modules/ Hopefully Pinecone has something in the works as well, and in the future tests can be made much simpler and more robust.

**Comment by neon-sunset** [comment]
@spppaul @maumar 2.0 version has been pushed to Nuget https://github.com/neon-sunset/Pinecone.NET/releases/tag/2.0.0

Let me know if something breaks.
Thanks!

**Comment by spppaul** [comment]
> @spppaul @maumar 2.0 version has been pushed to Nuget https://www.nuget.org/packages/Pinecone.NET/2.0.0
> 
> Let me know if something breaks. Thanks!

I removed my local reference and used the 2.0 nuget package and it seemed to work the same, thanks!

**Pull request merged**
